cuqi/experimental/algebra/_ast.py

```diff
@@ -0,0 +1,315 @@
+        return DivideNode(self, convert_to_node(other))
+
+    def __rtruediv__(self, other):
+        return DivideNode(convert_to_node(other), self)
+
+    def __pow__(self, other):
+        return PowerNode(self, convert_to_node(other))
+
+    def __rpow__(self, other):
+        return PowerNode(convert_to_node(other), self)
+
+    def __neg__(self):
+        return NegateNode(self)
+
+    def __abs__(self):
+        return AbsNode(self)
+
+    def __getitem__(self, i):
+        return GetItemNode(self, convert_to_node(i))
+
+    def __matmul__(self, other):
+        return MatMulNode(self, convert_to_node(other))
+
+    def __rmatmul__(self, other):
+        return MatMulNode(convert_to_node(other), self)
+
+
+class UnaryNode(Node, ABC):
+    """Base class for all unary nodes in the abstract syntax tree.
+
+    Parameters
+    ----------
+    child : Node
+        The direct child node on which the unary operation is performed.
+
+    """
+
+    def __init__(self, child: Node):
+        self.child = child
+
+
+class BinaryNode(Node, ABC):
+    """Base class for all binary nodes in the abstract syntax tree.
+
+    The op_symbol attribute is used for printing the operation in the __repr__ method.
+
+    Parameters
+    ----------
+    left : Node
+        Left child node to the binary operation.
+
+    right : Node
+        Right child node to the binary operation.
+
+    """
+
+    @property
+    @abstractmethod
+    def op_symbol(self):
+        """Symbol used to represent the operation in the __repr__ method."""
+        pass
+
+    def __init__(self, left: Node, right: Node):
+        self.left = left
+        self.right = right
+
+    def __repr__(self):
+        return f"{self.left} {self.op_symbol} {self.right}"
+
+
+class BinaryNodeWithParenthesis(BinaryNode, ABC):
+    """Base class for all binary nodes in the abstract syntax tree that should be printed with parenthesis."""
+
+    def __repr__(self):
+        left = f"({self.left})" if isinstance(self.left, BinaryNode) else str(self.left)
+        right = (
+            f"({self.right})" if isinstance(self.right, BinaryNode) else str(self.right)
+        )
+        return f"{left} {self.op_symbol} {right}"
+
+
+# ====== Specific implementations of the "leaf" nodes ======
+
+
+class VariableNode(Node):
+    """Node that represents a generic variable, e.g. "x" or "y".
+
+    Parameters
+    ----------
+    name : str
+        Name of the variable. Used for printing and to retrieve the given input value
+        of the variable in the kwargs dictionary when evaluating the tree.
+
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    def __call__(self, **kwargs):
+        """Retrieves the value of the variable from the passed kwargs. If no value is found, it raises a KeyError."""
+        if not self.name in kwargs:
+            raise KeyError(
+                f"Variable '{self.name}' not found in the given input parameters. Unable to evaluate the expression."
```

Comment:
do you think catching unmatched keyword arguments should generate and error? for example:

`(x+1)(x=1, y=3)`

Reply:
I think that would be good, although I am not sure how to handle that currently using this "localized" structure. In the random variables we can catch it, but here I am not sure how to do it.

Reply:
I added the point to random variables. I think it is better dealt with there after we look into that task

Reply:
#577

Reply:
Sounds good plan!